gflownet/envs/base.py

```diff
@@ -207,19 +218,51 @@ def step(self, action: Tuple[int]) -> Tuple[List[int], Tuple[int], bool]:
             False, if the action is not allowed for the current state, e.g. stop at the
             root state
         """
-        # If env is done, return invalid
-        if self.done:
-            return self.state, action, False
         # If action not found in action space raise an error
         if action not in self.action_space:
             raise ValueError(
                 f"Tried to execute action {action} not present in action space."
             )
-        action_idx = self.action_space.index(action)
+        # If env is done, return invalid
+        if self.done:
+            return False, self.state, action, False
         # If action is in invalid mask, exit immediately
-        if self.get_mask_invalid_actions_forward()[action_idx]:
-            return self.state, action, False
-        return None, None, None
+        if not (self.skip_mask_check or skip_mask_check):
+            action_idx = self.action_space.index(action)
+            if self.get_mask_invalid_actions_forward()[action_idx]:
+                return False, self.state, action, False
+        return True, self.state, action, True
```

Comment:
In all of this code, the outputs `do_step` and `valid` are always identical. Is there a reason to have both outputs?

Reply:
I thought of that too. The main reason is that in the future `_pre_step()` could be extended (or overwritten by certain environments) in ways that could yield a result with different `do_step` and `valid`. Another reason is that semantically they are different.

Reply:
Right, the only example I can think of is when the action is EOS. The action will be valid but _there is no need to proceed further_. Anyway, you have almost convinced me that it might be unnecessary to have both outputs. Should we keep only `do_step`?

Reply:
Ok, changed [here](https://github.com/alexhernandezgarcia/gflownet/pull/138/commits/c5011f35d20a601c7806b2f9693051a58e84c629)